### PR TITLE
[Sequence] Add "intermediate" methods

### PR DESCRIPTION
**Pull request by nikophil**
Adds the lazy stage of Sequence: filterNotNull, filterInstanceOf, mapNotNull, flatMap, flatten, takeFirst, dropFirst, takeWhile, dropWhile, distinct, distinctBy, zip, zipWithNext and onEach. Each returns a new Sequence and pulls nothing until a terminal operation runs.

ZipOperation stops buffering the other side into an array and walks both sides in lockstep, reading no further than the shorter one needs. This changes Collection::zip() too, since the operation is shared. A Generator resumes where it stands, every other iterator is rewound and so restarts, and a NoRewindIterator opts out of that. Replayability is read off the same rule, so a second pass is refused only for cursors that were never rewound.

SequenceAlreadyIteratedException becomes NonReplayableSourceException, since the thing already iterated is not always the sequence. Also: takeFirst(n) now pulls exactly n elements instead of n+1, and the Sequence terminals carry @throws tags.

Refs #23

### CODE REVIEW

File: src/Sequence/Sequence.php [hunk 1]
```diff
@@ -66,6 +84,132 @@ public function filter(Closure $predicate): Sequence;
...
+	 *
+	 * The lazy, chainable counterpart of forEach: the action runs once per element and per
+	 * pass, while the element flows through the pipeline - nothing happens before a
+	 * terminal operation pulls.
+	 *
+	 * @param Closure(E, int):void $action
+	 * @return Sequence<E>
+	 */
+	#[NoDiscard]
+	public function onEach(Closure $action): Sequence;
```
Comment (nikophil) [inline src/Sequence/Sequence.php:219]: I'm introducing `onEach(): Sequence` before `forEach(): void`

there is currently no BC break

File: src/Operation/ZipOperation.php [hunk 2]
```diff
@@ -26,17 +28,50 @@ final class ZipOperation extends AbstractOperation
...
+	 * @return Iterator<T>
+	 */
+	private static function cursor(iterable $iterable): Iterator
+	{
+		if (is_array($iterable)) {
+			return new ArrayIterator($iterable);
+		}
+
+		if ($iterable instanceof Iterator) {
+			return $iterable;
```
Comment (delacry) [inline src/Operation/ZipOperation.php]: shouldn't this also `rewind()` when possible? like:
```php
try {
	$iterable->rewind();
} catch (Throwable) {
	// A cursor that has already started: keep it where it is.
}
```
but probably not the try-catch.. just
```php
$iterable->rewind();
```

I'm not leaning towards try-catch, because `sequenceOf(started generator)` just throws, so you cannot create sequence from generator that's already started.. but this would work `sequenceOf([1,2])->zip($startedGen)`, and is that expected to work?

I've checked how it works in Kotlin, and it doesn't do `rewind()` at all, it continues from current position of the started generator, that works because Kotlin's Iterator is always positioned - `hasNext()` can be used the moment you get it.. in PHP a fresh iterator and half-consumed one are indistinguishable, both are just cursors sitting somewhere, so if we skip rewind to support started case, `zip(new SplStack)` silently returns `[]`, we can detect it for generators (they throw), but not for `ArrayIterator`/`SplStack`

here's maybe better explanation by AI, because I'm not sure I got it right:
```
  the stack has two elements. That's exactly the bug:

  count($stack)           = 2
  foreach gives           = b a
  valid() before rewind() = false      ← the trap
  zip($stack)             = []

  SplStack (via SplDoublyLinkedList) doesn't position its internal pointer until rewind() is called. foreach calls rewind() for you, so iterating it normally yields both elements. cursor()
  skips the rewind, so it asks valid() on an unpositioned list, gets false, and the while ($left->valid() && $right->valid()) loop never runs a single iteration.

  So zip() silently reports "the other side is empty" about a list that plainly isn't. Nothing throws, nothing warns — you just get fewer pairs than you have data, or none at all.

  That's the difference between the two states I described:

  - valid() === false because it's exhausted → [] is correct
  - valid() === false because it was never rewound → [] is wrong

  Both look identical to cursor(). rewind() is what collapses the ambiguity: after rewinding, valid() === false reliably means "actually empty."

  Fix your comment's wording — zip(new SplStack) does read like an empty stack. Better:

  $stack = new SplStack();
  $stack->push('a');
  $stack->push('b');

  count($stack);                            // 2
  iterator_to_array($stack);                // ['b', 'a']
  listOf([1, 2])->zip($stack)->toArray();   // 0.2.x: [[1,'b'], [2,'a']]
                                            // here:  []

  The count() and iterator_to_array() lines are worth keeping in — they make it obvious the data is there and that the empty result is a bug rather than an edge case about empty inputs.
```

File: tests/Sequence/SequenceIterationTest.php [hunk 3]
```diff
@@ -142,6 +142,85 @@ public function a_consumed_iterator_is_not_retained_by_the_sequence(): void
...
+		$this->assertSame([[1, 'a'], [2, 'b']], $sequence->toArray());
+
+		// Without the guard the other side would resume where it stopped and the second pass
+		// would silently pair 1 with 'c' - the failure mode the contract exists to prevent.
+		$this->expectException(SequenceAlreadyIteratedException::class);
+		$this->expectExceptionMessageIsOrContains(
+			'This sequence is backed by a non-replayable source and has already been iterated. Create a new sequence from a fresh source to iterate again.',
+		);
+
+        // phpcs:ignore
```
Comment (delacry) [inline tests/Sequence/SequenceIterationTest.php]: I see that some spaces have been smuggled 😄 i missed it in #24, it's there at the similar places

File: src/Sequence/SequenceLogic.php [hunk 4]
```diff
@@ -61,20 +68,143 @@ public function getIterator(): Generator
...
+	{
+		$otherIsOneShot = $other instanceof Traversable && !$other instanceof IteratorAggregate;
+		$otherConsumed = false;
+
+		return $this->newSequenceOf(function () use ($other, $otherIsOneShot, &$otherConsumed): iterable {
+			if ($otherIsOneShot && $otherConsumed) {
+				throw SequenceAlreadyIteratedException::nonReplayableSourceAlreadyIterated();
+			}
+
+			$otherConsumed = true;
```
Comment (delacry) [inline src/Sequence/SequenceLogic.php]: this flips even though `ZipOperation::with()` didn't pull anything, if you run `zip()` on empty sequence on the left with some generator on the right for example

```php
$gen = (function (): Generator { yield 'a'; yield 'b'; })();
$z = sequenceOf([])->zip($gen);

$z->toArray(); // []
$z->toArray(); // SequenceAlreadyIteratedException, but $gen is still at "a"
```

same could be done with `->takeFirst(0)->zip($gen)` - nothing was consumed, so the second pass would have worked fine, so we should wrap it like:

```php
return (function () use ($other, &$otherConsumed): Generator {
	foreach (new ZipOperation($this)->with($other) as $pair) {
		$otherConsumed = true;

		yield $pair;
	}
})();
```

also some test for this behavior would be cool

File: src/Sequence/SequenceLogic.php [hunk 5]
```diff
@@ -61,20 +68,143 @@ public function getIterator(): Generator
...
+	/** {@inheritDoc} */
+	#[NoDiscard]
+	public function zip(iterable $other): Sequence
+	{
+		$otherIsOneShot = $other instanceof Traversable && !$other instanceof IteratorAggregate;
+		$otherConsumed = false;
+
+		return $this->newSequenceOf(function () use ($other, $otherIsOneShot, &$otherConsumed): iterable {
+			if ($otherIsOneShot && $otherConsumed) {
+				throw SequenceAlreadyIteratedException::nonReplayableSourceAlreadyIterated();
```
Comment (delacry) [inline src/Sequence/SequenceLogic.php]: this exception message might be misleading, it doesn't say which source, user might think it's the one you are calling `zip()` on

**Review by delacry — CHANGES_REQUESTED** [review]
Thanks, there are just few things we need to get right, and also I noticed we are missing `@throws` on `Sequence` interface, but maybe that's intended until we know all the exceptions from all possible methods.

File: src/Operation/ZipOperation.php (hunk 2 above)
Comment (delacry) [reply src/Operation/ZipOperation.php]: > because sequenceOf(started generator) just throws

just a correction, it throws probably later, when we try to do foreach or some terminal op, because that calls `rewind()` and `Generator` rejects that.. maybe this could be somehow prevented

I've also inspected it deeper, and found out that `Generator` is fine, the problem is just with non-Generator Iterators where `valid()` is meanlingless until `rewind()`.. so probably we could support the never-rewind for `Generator`, so it matches Kotlin behavior.

File: tests/Sequence/SequenceIterationTest.php (hunk 3 above)
Comment (nikophil) [reply tests/Sequence/SequenceIterationTest.php]: haha my bad :sweat_smile: 

I'll add the specific rule to ignore

File: src/Sequence/SequenceLogic.php (hunk 5 above)
Comment (nikophil) [reply src/Sequence/SequenceLogic.php]: ok, I changed the exception class name to `NonReplayableSourceException` and added a named ctor.

BTW, don't you think all that stuff that protects the the `zip()` parameter should be added directly in `ZipOperation`?

File: src/Sequence/SequenceLogic.php (hunk 4 above)
Comment (nikophil) [reply src/Sequence/SequenceLogic.php]: fixed and test added

File: src/Operation/ZipOperation.php (hunk 2 above)
Comment (nikophil) [reply src/Operation/ZipOperation.php]: nice catch, I've added the `rewind()` and added some non regression tests

**Comment by nikophil** [comment]
> I noticed we are missing @throws on Sequence interface, but maybe that's intended until we know all the exceptions from all possible methods.

nope this was not intended, I've added the `@throws` tags

File: src/Operation/ZipOperation.php (hunk 2 above)
Comment (delacry) [reply src/Operation/ZipOperation.php]: so we will not support continuing the generator that's already being walked? because in my last message I concluded that it applies to non-Generator Iterators where that valid() problem occurs, for Generator it should be fine, so I'm not sure if we should skip rewind() on generators as it's in Kotlin, but rewind all other iterators..

Comment (nikophil) [reply src/Operation/ZipOperation.php]: hum sorry I just missed this. I've added an exception for Generators

File: tests/Sequence/SequenceIterationTest.php (hunk 3 above)
Comment (delacry) [reply tests/Sequence/SequenceIterationTest.php]: btw the spaces are still there, at each `// phpcs:ignore`, it's no big deal though, it probably doesn't flag comment-only lines, but it's probably worth getting right

File: src/Operation/ZipOperation.php (hunk 2 above)
Comment (delacry) [reply src/Operation/ZipOperation.php]: thanks, maybe it's worth adding some tip to the api docs under Collection::zip() like:
```
::: tip Resuming an advanced iterator
Only a `Generator` can be distinguished from a fresh cursor — every other advanced iterator is indistinguishable from an unstarted one and gets rewound. Wrap it in a `NoRewindIterator` to resume it where it stands instead.
:::
```

File: src/Sequence/SequenceLogic.php (hunk 5 above)
Comment (delacry) [reply src/Sequence/SequenceLogic.php]: yeah I think so, because we are deciding the same thing at 2 places, and probably not consistentnly.. because `ZipOperation::cursor()` is deciding which cursors get rewound and actually knows which ones are replayable, and here that `zip()` is doing the same with `$other instanceof Traversable && !$other instanceof IteratorAggregate`, they agree on Generator but disagree on everything else

so I would move the classification into `ZipOperation`, next to `cursor()`, like `isReplayable(iterable $other): bool`, or maybe if we are using it at other places it might even be some `@internal` method in `functions.php`, but the cross-pass tracking should be kept in `SequenceLogic`, it's a sequence-only state.

Comment (delacry) [reply src/Sequence/SequenceLogic.php]: and also when we are rewinding, maybe we should always wrap it so we throw ours exception instead of plain PHP, like:
```php
try {
	$cursor->rewind();
} catch (Throwable $e) {
	throw NonReplayableSourceException::zippedIterableCannotRewind($e);
}
  ```

File: src/Operation/ZipOperation.php (hunk 2 above)
Comment (nikophil) [reply src/Operation/ZipOperation.php]: done

File: tests/Sequence/SequenceIterationTest.php (hunk 3 above)
Comment (nikophil) [reply tests/Sequence/SequenceIterationTest.php]: lol sorry. that's strange that cbf did not fix it, nor PHPStorm with `.editorconfig` file

anyway, it's fixed now

File: src/Sequence/SequenceLogic.php (hunk 5 above)
Comment (nikophil) [reply src/Sequence/SequenceLogic.php]: done. I'm just not super happy with the error message, we don't know if the error occurred from the "other" or from the source

Comment (delacry) [reply src/Sequence/SequenceLogic.php]: yeah, that's maybe alright in that case - the left side is probably always safe anyways, and it's very unlikely someone would need to distinguish it in code, but feel free to improve if it can be done without complicating code too much, maybe in the next PR, I'm going to merge this one

Comment (nikophil) [reply src/Sequence/SequenceLogic.php]: yeah I also don't think the left side could break

**Review by delacry — APPROVED** [review]
Great work, thank you! Btw I squash-merge PRs, so feel free to put whatever you want in the PR description in next PRs, it will end up as the commit body... if it's not there I usually generate that short summary by AI, it's later helpful for searching things and also more context for agents.
